src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -85,13 +91,67 @@ class DefaultSource
     new SequenceRowsRelation(config, sequenceId)(sqlContext)
   }
 
+  private def getFunctionResult(client: GenericClient[IO], functionId: Long, result: FunctionCall): IO[Json] = {
+    var res = result
+    while (res.status != None && res.status.get == "Running") {
+      res = client
+        .functionCalls(functionId)
+        .retrieveById(res.id.get)
+        .unsafeRunSync()
+    }
+
+    if (res.status != None && res.status.get == "Completed") {
+      for {
+        functionCallResponse <- client
+          .functionCalls(functionId)
+          .retrieveResponse(res.id.getOrElse(0L))
+      } yield functionCallResponse.response.getOrElse(Json.fromJsonObject(JsonObject.empty))
+    } else {
+      IO.pure(Json.fromJsonObject(JsonObject.empty))
+    }
+  }
+
+  private def callFunctionAndGetResult(config: RelationConfig, functionId: Long, data: Json): IO[Json] = {
+    implicit val backend: SttpBackend[IO, Nothing] = CdpConnector.retryingSttpBackend(config.maxRetries)
+    for {
+      client <- GenericClient.forAuth[IO](Constants.SparkDatasourceVersion, config.functionsAuth.get, config.baseUrl, apiVersion = Some("playground"))
+      call <- client.functionCalls(functionId).callFunction(Json.fromJsonObject(JsonObject.empty))
+      res <- getFunctionResult(client, functionId, call)
+    } yield res
+  }
+
+  private def initializeFunctionUdfs(config: RelationConfig, context: SQLContext): Unit = {
+    implicit val backend: SttpBackend[IO, Nothing] = CdpConnector.retryingSttpBackend(config.maxRetries)
+
+    val functionsIO = for {
+      client <- GenericClient.forAuth[IO](Constants.SparkDatasourceVersion, config.functionsAuth.get, config.baseUrl, apiVersion = Some("playground"))
+      functions <- client.functions.read()
+      l <- client.login.status()
+      f = functions.items.map(func =>
+        (func.name.replace(" ", "_"), (data: Any) =>
```

Comment:
We'd need some prefix to make this work, but this also assumes that the user has permissions to list CDF functions. I think the benefit is small compared to having a `cdf_function("my function", <args>)` (using name/externalId), `cdf_function(123, <args>)` (using id), and it has some downsides.

Any particular reason why we'd like to do it this way?

Reply:
I just assumed that if a user has permissions to call a function, then he probably has permissions to list functions. Not sure if those are separate permissions, but if so, then we can refactor it.

Reply:
Still, I don't think registering all of them automatically is something we want to do: risk of collisions and possibly a performance hit if there are many. Doesn't seem worth it over `cdf_function("my function", <args>)` imo.